thumbnails/Registry.cpp

```diff
@@ -0,0 +1,123 @@
+#include "Registry.hpp"
+
+const std::string ThumbnailProviderRegistration::thumbnailProviderCLSID_ = "{e357fccd-a995-4576-b01f-234630154e96}";
+
+template<bool throwOnError>
+Registry<throwOnError>::Registry(std::string const & subKey, CreateMode createMode, HKEY root)
+{
+  if (createMode == CreateMode::Create)
+  {
+    check(
+      RegCreateKeyExA(
+        root, subKey.size() ? subKey.c_str() : nullptr, 0, nullptr, REG_OPTION_NON_VOLATILE, KEY_READ | KEY_WRITE | DELETE,
+        nullptr, &hKey_, nullptr),
+      "Error creating key: " + subKey);
+  }
+  else
+  {
+    check(
+      RegOpenKeyExA(root, subKey.size() ? subKey.c_str() : nullptr, 0, KEY_READ | KEY_WRITE | DELETE, &hKey_),
+      "Error opening key: " + subKey);
+  }
+}
+
+template<bool throwOnError>
+Registry<throwOnError>::~Registry()
+{
+  if (hKey_) {
+    check(RegCloseKey(hKey_), "Error closing key");
+  }
+}
+
+template<bool throwOnError>
+std::string Registry<throwOnError>::readString(std::string const & valueName)
+{ 
+  DWORD buflen = MAX_PATH-1;
+  DWORD type;
+  uint8_t buf[MAX_PATH] = {};
+  LSTATUS result = RegQueryValueExA(hKey_, valueName.c_str(), nullptr,
+    &type, buf, &buflen);
+  
+  if (valueName.size() == 0 && (result == ERROR_FILE_NOT_FOUND))
+    return "";
+
+  check(result,
+    "Error reading value");
+  if (type != REG_SZ && type != REG_MULTI_SZ && type != REG_EXPAND_SZ)
+    throw std::runtime_error("Value has unexpected type.");
```

Comment:
Throws even if throwOnError is false which looks fishy. It seems like throwOnError=false isn't really used at all except to check if a key exists? How about a static Registry::keyExists(...) method and removing the throwOnError template stuff?